File: soap/Q2252_CharacterMaskSortBySeverity.java
import java.util.ArrayList;
import java.util.HashMap;
import java.util.List;
import java.util.Map;
import soap.q2221.generated.CharacterService;
import soap.q2221.generated.SoapCharacterService;

public class Q2252_CharacterMaskSortBySeverity {
    static final String STUDENT_CODE = "B21DCCN001";   // TODO
    static final String Q_CODE       = "TODO_qCode";   // TODO

    static final String EMAIL_RE = "[A-Za-z0-9._%+-]+@[A-Za-z0-9.-]+\\.[A-Za-z]{2,}";
    static final String PHONE_RE = "(?<!\\d)0\\d{9}(?!\\d)";
    static final String TOKEN_RE = "token=\\S+";

    static final Map<String, Integer> SEVERITY = new HashMap<>();
    static {
        SEVERITY.put("ERROR", 0);
        SEVERITY.put("WARN",  1);
        SEVERITY.put("INFO",  2);
    }

    public static void main(String[] args) {
        CharacterService service = new CharacterService();
        SoapCharacterService port = service.getSoapCharacterServicePort();

        List<String> logs = port.requestStringArray(STUDENT_CODE, Q_CODE);
        System.out.println("logs(in)=" + logs.size() + " dòng");

        // 1) che dữ liệu nhạy cảm
        List<String> redacted = new ArrayList<>(logs.size());
        for (String line : logs) {
            redacted.add(line
                    .replaceAll(EMAIL_RE, "[EMAIL]")
                    .replaceAll(PHONE_RE, "[PHONE]")
                    .replaceAll(TOKEN_RE, "token=[TOKEN]"));
        }

        // 2) sort theo mức ERROR < WARN < INFO (token đầu tiên); ổn định (stable sort)
        redacted.sort((a, b) -> {
            int sa = SEVERITY.getOrDefault(firstToken(a), Integer.MAX_VALUE);
            int sb = SEVERITY.getOrDefault(firstToken(b), Integer.MAX_VALUE);
            return Integer.compare(sa, sb);
        });
        System.out.println("logs(sorted)=" + redacted);

        port.submitStringArray(STUDENT_CODE, Q_CODE, redacted);
        System.out.println("Đã submit " + redacted.size() + " dòng");
    }

    static String firstToken(String s) {
        int sp = s.indexOf(' ');
        return sp < 0 ? s : s.substring(0, sp);
    }
}

File: soap/Q2252_CharacterMaskSortBySeverity.py
import re
from zeep import Client

EXAM_IP      = "36.50.135.242"
STUDENT_CODE = "B22DCCN863"   # TODO
Q_CODE       = "TODO_qCode"   # TODO
WSDL         = f"http://{EXAM_IP}:2221/CharacterService?wsdl"

EMAIL_RE = re.compile(r"[A-Za-z0-9._%+-]+@[A-Za-z0-9.-]+\.[A-Za-z]{2,}")
PHONE_RE = re.compile(r"(?<!\d)0\d{9}(?!\d)")
TOKEN_RE = re.compile(r"token=\S+")

SEVERITY = {"ERROR": 0, "WARN": 1, "INFO": 2}


def redact(line: str) -> str:
    line = EMAIL_RE.sub("[EMAIL]", line)
    line = PHONE_RE.sub("[PHONE]", line)
    line = TOKEN_RE.sub("token=[TOKEN]", line)
    return line


def severity_key(line: str) -> int:
    first = line.split(" ", 1)[0] if " " in line else line
    return SEVERITY.get(first, 10**9)


def main():
    client = Client(WSDL)

    logs = client.service.requestStringArray(STUDENT_CODE, Q_CODE)
    print(f"logs(in)= {len(logs)} dòng")

    redacted = [redact(line) for line in logs]
    redacted.sort(key=severity_key)   # Python sort ổn định
    print("logs(sorted)=", redacted)

    client.service.submitStringArray(STUDENT_CODE, Q_CODE, redacted)
    print(f"Đã submit {len(redacted)} dòng")


if __name__ == "__main__":
    main()
